akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala

```diff
@@ -347,10 +350,40 @@ private[stream] class OutgoingConnectionStage(
       manager,
       Connect(remoteAddress, localAddress, options, connTimeout, pullMode = true),
       localAddressPromise,
-      halfClose))
+      halfClose),
+      remoteAddress)
 
     (logic, localAddressPromise.future.map(OutgoingConnection(remoteAddress, _))(ExecutionContexts.sameThreadExecutionContext))
   }
 
   override def toString = s"TCP-to($remoteAddress)"
 }
+
+/** INTERNAL API */
+private[akka] object TcpIdleTimeout {
+  def apply(idleTimeout: FiniteDuration, remoteAddress: Option[InetSocketAddress]): BidiFlow[ByteString, ByteString, ByteString, ByteString, NotUsed] = {
+    val connectionToString = remoteAddress match {
+      case Some(addr) ⇒ s" on connection to [$addr]"
+      case _          ⇒ ""
+    }
+
+    val toNetTimeout: BidiFlow[ByteString, ByteString, ByteString, ByteString, NotUsed] =
+      BidiFlow.fromFlows(
+        Flow[ByteString].mapError {
+          case t: TimeoutException ⇒
+            new TcpIdleTimeoutException(s"TCP idle-timeout encountered$connectionToString, no bytes passed in the last $idleTimeout", idleTimeout)
+        },
+        Flow[ByteString]
+      )
+    val fromNetTimeout: BidiFlow[ByteString, ByteString, ByteString, ByteString, NotUsed] =
+      BidiFlow.fromFlows(
+        Flow[ByteString],
+        Flow[ByteString].mapError {
```

Comment:
this flow could also be reused, but I think you can just reuse the previous bidiFlow reversed.